### ISSUE TITLE
Enable import functionality for any content on all websites

### ISSUE THREAD

**Issue by Haini** [JabRef/JabRef-Browser-Extension#175]
Current state: Can't import any PDF / Website into my references, because we are only considering pages that possibly contain reference data (?). 
![image](https://user-images.githubusercontent.com/5941765/82820867-47390b00-9ea3-11ea-8668-ac26da50a6a6.png)

Desired state: We should be able to import any PDF / Website, even if there is no metadata available. The user can then edit the fields for the new entry to his liking while the PDF / HTML is saved to the set target folder.

This is essentially what the Zotero Plugin already can, and as we are using Zotero at the core of this plugin anyways it makes sense to support this functionality.

**Comment by tobiasdiez** [comment]
So, just push the URL to JabRef without any other metadata?

**Comment by Haini** [comment]
Exactly this behavior would be great. 

Any reason why we shouldn't have this behavior?

**Comment by tobiasdiez** [comment]
I think this would indeed be a valuable addition! Please feel free to dive deeper! Here are a few pointers:
- https://github.com/JabRef/JabRef-Browser-Extension/blob/734296660dadc4dbe99b4b4d5b17d7dad7eabb32/background.js#L72 needs to be replaced by something like https://github.com/JabRef/JabRef-Browser-Extension/blob/734296660dadc4dbe99b4b4d5b17d7dad7eabb32/background.js#L101-L105 in order to show the JabRef icon
- https://github.com/JabRef/JabRef-Browser-Extension/blob/734296660dadc4dbe99b4b4d5b17d7dad7eabb32/background.js#L122 needs to be changed to invoke `saveAsWebpage` as in
https://github.com/zotero/zotero-connectors/blob/9fd1ac22cd0cc4839e782ed9082e7492c5ef7532/src/browserExt/background.js#L655
https://github.com/zotero/zotero-connectors/blob/9fd1ac22cd0cc4839e782ed9082e7492c5ef7532/src/browserExt/background.js#L683
- If we are lucky, then everything works already. But probably this is not the case ;-) This is how the flow should go (if I'm not mistaken): 
https://github.com/JabRef/JabRef-Browser-Extension/blob/fc3ab66d8a6788d2e1a408a95c74d87555349e00/Zotero/inject.js#L532
https://github.com/JabRef/JabRef-Browser-Extension/blob/afed0dc2c427014a6ee77237de2b983f23830b23/Zotero/translate_item.js#L75
https://github.com/JabRef/JabRef-Browser-Extension/blob/afed0dc2c427014a6ee77237de2b983f23830b23/Zotero/translate_item.js#L94
https://github.com/JabRef/JabRef-Browser-Extension/blob/734296660dadc4dbe99b4b4d5b17d7dad7eabb32/connector.js#L15-L17

**Comment by Haini** [comment]
tl;dr: I played around, a proof of concept works but there is a LOT of work to do for a proper and sound solution as other functionality breaks too. Also, currently only tested for Firefox. 

---

Hi and sorry for the late answer. I played around with this a little bit, and of course it doesn't just work out of the box. 
The following problems seem to arise:
- Firefox does not allow listeners in PDF Tabs, so there is a special function for this, which does not work in the default setup (It says that `Zotero.Utilities.saveFirefoxPDF` is not a function, I suppose the corresponding file is just not loaded?) 

https://github.com/Haini/JabRef-Browser-Extension/blob/d47f097bb93a92fcd643d5a69dca981da4dab032/zotero-connectors/src/browserExt/firefoxPDF.js#L50-L53

I went ahead and hacked the (seemingly) relevant part of the code in. Also note that `Zotero.browserMajorVersion` is undefined as well as `_tabInfo`, so I removed them for now from the query: 

https://github.com/Haini/JabRef-Browser-Extension/blob/d47f097bb93a92fcd643d5a69dca981da4dab032/Zotero/background.js#L864-L887

Also note, that the catch-all function runs into nothingness (at least under firefox), i.e. there is no one that could listen for the message and I don't fully understand the Zotero message magic yet:

https://github.com/Haini/JabRef-Browser-Extension/blob/d47f097bb93a92fcd643d5a69dca981da4dab032/Zotero/background.js#L889-L891

---

Next up for the connector: 

https://github.com/Haini/JabRef-Browser-Extension/blob/d47f097bb93a92fcd643d5a69dca981da4dab032/connector.js#L19-L36

https://github.com/Haini/JabRef-Browser-Extension/blob/d47f097bb93a92fcd643d5a69dca981da4dab032/connector.js#L156

```
				console.log("Haini: Trying to send to Jabref with tabId %s, %s,  %o", tab.id, tab.url, items)
				return browser.tabs.sendMessage(
					2, { // WORKAROUND: This 2 should a tab.id
						convertToBibTex: items
```

I just hacked together a proof of concept for this one. I am basically creating a dummy entry that contains the relevant information (Url, Tab Title, Attachment) and then send it to another tab where hopefully a translator listens (as there is no listener in the PDF tab). 
I hope that I just overlooked something and that there is an elegant solution.

---

Next step would be to identify a proper solution for the "no listener in PDF Tabs" problem. But I don't know enough about how to handle bigger Javascript projects and why e.g. `Zotero.Utilities.saveFirefoxPDF` is not already loaded.

**Comment by tobiasdiez** [comment]
Thanks for investigating! This looks already pretty promising. I suggest you open a new PR and we continue the discussion there.

Just a quick suggestion after reading a bit about the limitations content scripts for pdfs in Firefox: what about circumventing the whole translation business and construct the relevant bibtex code by hand? If a PDF is shown, then there is no metadata anyway to extract so once you have the url, you have all information and can send
```
@misc{unknown,
file={PDF:<url>:pdf},
url = {<url>}
}
```
directly to JabRef. This can be done easily from the background script and you don't need to modify/call any Zotero functions. What do you think?

**Comment by Haini** [comment]
Thanks for the feedback, I will do so as soon as I have time :) 

Wrt to your proposal: while manually constructing an object I also thought about this and I see no reason why we shouldn't just implement it that way! I just wanted to keep it a little bit more general (and I was trying to understand how the messaging system and your listeners work, so I had to experiment a little bit anyways). 

I will try to change the behavior accordingly and open a pull request for further discussion :+1:

**Comment by tobiasdiez** [comment]
Thanks for your suggestion! With the new version, it should be possible now to directly import PDFs into JabRef.
